sycl/doc/extensions/proposed/sycl_ext_intel_queue_index.asciidoc

```diff
@@ -0,0 +1,194 @@
+SYCL specification refer to that revision.
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Overview
+
+Backends such as Level Zero and OpenCL expose an "index" to a device's work
+submission queue, which allows the application to fine tune the way work is
+submitted to a device.  This extension exposes that same concept to SYCL
+applications.
+
+Most SYCL applications should not need to use this extension because the SYCL
+implementation automatically selects an efficient way to submit work to a
+device, including automatic selection of a queue index when necessary.
+Therefore, this extension is aimed at advanced users who understand the device
+hardware and think they can outperform the default implementation by specifying
+an explicit queue index.
+
+Note that this extension can be supported on any backend, even if the backend
+has no notion of a "queue index".  Backends that have no native support for a
+queue index can report that a device has only a single available queue index.
+Applications can then only request one possible queue index, and the backend
+can treat this as the default behavior (i.e. the backend can ignore the index).
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_INTEL_QUEUE_INDEX` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+=== New device information descriptor
+
+This extension adds the following new device information descriptor which
+allows the application to query the number of available queue indices for the
+device.
+
+```
+namespace sycl::ext::intel::info::device {
+
+struct max_compute_queue_indices;
+
+} // namespace sycl::ext::intel::info::device
+```
+
+The return type for this information descriptor is `int`, and the value is a
+positive integer telling the number queue indices that are available for the
+device.  These indices are numbered sequentially starting at `0`.
+
+=== New queue property
+
+This extension adds the following new queue property which can be specified to
+the queue constructor via the `property_list` parameter.
+
+```
+namespace sycl::ext::intel::property::queue {
+
+class compute_index {
```

Comment:
Level Zero and OpenCL drivers have 2 levels of exposing those indexes.
The first level is command queue groups, there can be multiple of those and they can have different charecteristics, i.e. you can have Compute command queue group and Copy Command Queue group.
Then inside of those command queue groups we can have multiple indexes representing engines with those capabilieis.

Here I see that everything if flattened to one level, how would this map to what Level Zero and OpenCL drivers offers?

Reply:
Thanks for clarification, what would happen if we would have multiple compute groups within single device?

Reply:
I guess it would depend on what the semantic difference was between the various compute groups.  If this happened, my first concern would be the behavior of a SYCL `queue` when there is *no* property.  Would this choose one of the various compute groups arbitrarily?

Can you give me an example of why Level Zero might expose multiple compute groups?  What would be the reason for doing this vs. exposing multiple indices all within the same compute group?

Reply:
One reason would be special command queue group for cooperative dispatch kernels.
Another example would be exposing RCS/CCS separately.

Reply:
> One reason would be special command queue group for cooperative dispatch kernels.

In a case like this, I think it would be better to expose "cooperative dispatch" as a separate queue property, and we would make the following changes:

* Add a new queue property like `ext::intel::property::queue::cooperative_dispatch`.
* Add some device query to tell whether cooperative dispatch is supported.
* Assuming there were multiple queue indices for a cooperative dispatch queue:
  * Add a new device query like `ext::intel::info::device::max_cooperative_compute_queue_indices`
  * When the existing `compute_index` property is used together with the new `cooperative_dispatch` property, the index is interpreted as cooperative compute queue index.
* Clarify that the existing `ext::intel::info::device::max_compute_queue_indices` query tells the number of compute queue indices for normal (non-cooperative) queues.

I would use this same pattern for any other "special" compute command queue group that has some new semantic.

> Another example would be exposing RCS/CCS separately.

Sorry, "RCS" is a new acronym to me, so I can't answer how this would be reflected in SYCL.

Reply:
RCS is Render Command Streamer. CCS is compute command streamer.
RCS can run work as well as CCS.

Reply:
Is there some advantage / disadvantage to submitting work to a CCS vs. an RCS?  What would the default behavior be for a SYCL queue that has no property?  Should it run only on a CCS, or should it choose either a CCS or an RCS arbitrarily?  If there is no particular advantage to using a CCS vs. an RCS, then we would probably flatten the index space in SYCL.  However, I don't see why Level Zero would bother exposing them separately if they can be used interchangeably.

Reply:
@gmlueck RCS and CCS may have different advantages. Or going in a more generic approach: there could be different types of compute engines, in the same say we have different types of copy engines (main copy engine vs link engines). So we need to have something generic and flexible enough that could accommodate those kind of cases.

Reply:
I think I see @MichalMrozek concerns here. Here we are saying that there's a given number of indexes available

```
namespace sycl::ext::intel::info::device {

struct max_compute_queue_indices;

} /
```

however, not all are the same. From what L0 defines here https://spec.oneapi.io/level-zero/latest/core/PROG.html#discovery 

>The number and properties of command queue groups is queried by using [zeDeviceGetCommandQueueGroupProperties](https://spec.oneapi.io/level-zero/latest/core/api.html#zedevicegetcommandqueuegroupproperties).
>
>The number of physical engines within a group is queried from [ze_command_queue_group_properties_t](https://spec.oneapi.io/level-zero/latest/core/api.html#ze-command-queue-group-properties-t).numQueues.

so there's an amount of indices per queue group, while here we are presenting an aggregated count of those indices, which might not be completely correct. if we want to make this more generic, we should present also the notion of queue groups, so there's an amount of indices per group.

Reply:
The way I see it, the property name `compute_index` corresponds to the (first) "compute queue group".  If we want to expose other queue groups to SYCL in the future, we will add additional queue properties.  For example, if Level Zero decides to expose an RCS compute queue group, we can add a new SYCL queue property like `rcs_compute_index`.

I would rather *not* change `compute_index` to take both a group number and a queue index.  If we did that, we'd need to add documentation about what the group number means because it seems the different queue groups all have different semantics.  Since they have different semantics, I'd rather tie each to its own property, so that it's easy to document the semantics of each one.

Reply:
thanks @gmlueck . 

considering then that we are open to future additions or changes, then I have no immediate concerns.

+1

Reply:
Same from me +1